tests/utils.py

```diff
@@ -153,26 +148,46 @@ class TestServerProcess():
   """
 
 
-  def __init__(self, log, server='simple_server.py',
-      port=None, timeout=10, popen_cwd=".",
-      extra_cmd_args=[]):
+  def __init__(self, log, server='simple_server.py', timeout=10,
+      popen_cwd=".", extra_cmd_args=[]):
 
     # Create temporary log file used for logging stdout and stderr
     # of the subprocess. In the mode "r+"" stands for reading and writing
     # and "t" stands for text mode.
     self.__temp_log_file = tempfile.TemporaryFile(mode='r+t')
 
     self.server = server
-    self.port = port or random.randint(30000, 45000)
     self.__logger = log
 
-    # The "-u" option forces stdin, stdout and stderr to be unbuffered.
-    command = ['python', '-u', server, str(self.port)] + extra_cmd_args
-
-    # We are reusing one server subprocess in multiple unit tests, but we are
-    # collecting the logs per test.
-    self.__server_process = subprocess.Popen(command,
-        stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
+    started = False
+    port_generation_attempts = 5
+    while not started and port_generation_attempts > 0:
+      self.port = random.randint(30000, 45000)
+      # The "-u" option forces stdin, stdout and stderr to be unbuffered.
+      command = ['python', '-u', server, str(self.port)] + extra_cmd_args
+
+      # We are reusing one server subprocess in multiple unit tests, but we are
+      # collecting the logs per test.
+      self.__server_process = subprocess.Popen(command,
+          stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
+
+      # Some tests (like those in test_slow_retrieval_attack.py) require no
+      # checks if the server has started.
+      if timeout == 0:
+        break
+
+      started = self.has_server_started(timeout)
+      if not started:
+        # If the server has not started for whatever reason
+        self.__logger.info("Could not connect to the " + server + " on port " \
```

Comment:
```suggestion
        self.__logger.info("Failed to start " + server + " on port " \
```
We've not tried connecting